### ISSUE TITLE
Weird scrolling behavior

### ISSUE THREAD

**Issue by synaptiko** [jwilm/alacritty#1185]
I've noticed weird janks when scrolling from the beginning I use alacritty but I wasn't sure what are the reliable steps. Now I'm able to reproduce it on 100%. I'm using i3-wm so it can be related to combination of i3, tmux & alacritty. It doesn't happen in xfce4-terminal with tmux though.

Steps:
1. open alacritty on one workspace, open Chromium on another workspace
2. open tmux and vim inside tmux with some long file, scroll in the middle of the file
3. open some long page in Chromium
4. switch to alacritty, scroll up, scroll down, remember the line and notice that scrolling is working fine
5. switch to Chromium a scroll up for some time (few page, for example)
6. switch back to alacritty and trigger single scroll down or up and notice that scrolling move several lines/pages up

It seems to me that alacritty is not handling properly focus & blur events so when I scroll in another window in the system and get back to alacritty and trigger scroll it will replay all the scroll events which happened while alacritty was in the background (blurred).

My tmux config:
```
set -g default-terminal "tmux-256color"
set -ga terminal-overrides ",*256col*:Tc" # See https://github.com/tmux/tmux/issues/34
set -g mouse off
set -g history-limit 100000
set -s escape-time 0
set-option -g status off
set-window-option -g xterm-keys on

bind-key m set-option -g mouse off
bind-key M set-option -g mouse on

bind-key [ copy-mode \; send-keys 0 \; send-keys H

run-shell ~/.files/tmux/plugins/extrakto/extrakto.tmux
```

Arch Linux 4.15.8-1-ARCH x86_64 with X11
xorg-server 1.19.6
mesa 17.3.6
i3-wm 4.14.1-2
tmux 2.6

**Comment by NickeZ** [comment]
I experience a similar issue! I thought it was solved by all the mouse related patches recently but it wasn't. Great that you found a way to reproduce it!

**Comment by maximbaz** [comment]
Wow, I reproduce it too. 

Worth mentioning: it's best to switch between workspaces using keyboard shortcuts only, that is after step 5 don't touch the mouse, use keyboard to switch to Chrome and then make a scroll with the mouse. It reproduces if you switch workspaces with mouse too, but not every time, maybe a mouse click "resets" the scroll history sometimes.

**Comment by chrisduerr** [comment]
Can this be reproduced with the scrollback PR?

**Comment by maximbaz** [comment]
Yes, even with plain alacritty + nvim or man, no tmux needed.

**Comment by chrisduerr** [comment]
I have troubles reproducing this. But this might be something in `glutin`? I'd assume that the moment you scroll in alacritty again, we receive a large number of scroll events.

**Comment by maximbaz** [comment]
Possibly. Testing this: `$ RUST_LOG=trace ./target/debug/alacritty 2>&1 | grep scroll`

One smallest possible movement of a mouse wheel produces this:

```
TRACE 2018-03-14T18:14:19Z: alacritty::term: scroll_down_relative: origin=0, lines=1
```

Switch to Chromium, scroll a bit, switch back to alacritty and do one more smallest possible movement of a mouse wheel:

```
TRACE 2018-03-14T18:14:24Z: alacritty::term: scroll_down_relative: origin=0, lines=1
TRACE 2018-03-14T18:14:24Z: alacritty::term: scroll_down_relative: origin=0, lines=1
TRACE 2018-03-14T18:14:24Z: alacritty::term: scroll_down_relative: origin=0, lines=1
TRACE 2018-03-14T18:14:24Z: alacritty::term: scroll_down_relative: origin=0, lines=1
TRACE 2018-03-14T18:14:24Z: alacritty::term: scroll_down_relative: origin=0, lines=1
TRACE 2018-03-14T18:14:24Z: alacritty::term: scroll_down_relative: origin=0, lines=1
TRACE 2018-03-14T18:14:24Z: alacritty::term: scroll_down_relative: origin=0, lines=1
TRACE 2018-03-14T18:14:24Z: alacritty::term: scroll_down_relative: origin=0, lines=1
TRACE 2018-03-14T18:14:24Z: alacritty::term: scroll_down_relative: origin=0, lines=1
TRACE 2018-03-14T18:14:24Z: alacritty::term: scroll_down_relative: origin=0, lines=1
TRACE 2018-03-14T18:14:24Z: alacritty::term: scroll_down_relative: origin=0, lines=1
TRACE 2018-03-14T18:14:24Z: alacritty::term: scroll_down_relative: origin=0, lines=1
TRACE 2018-03-14T18:14:24Z: alacritty::term: scroll_down_relative: origin=0, lines=1
TRACE 2018-03-14T18:14:24Z: alacritty::term: scroll_down_relative: origin=0, lines=1
TRACE 2018-03-14T18:14:24Z: alacritty::term: scroll_down_relative: origin=0, lines=1
TRACE 2018-03-14T18:14:24Z: alacritty::term: scroll_down_relative: origin=0, lines=1
TRACE 2018-03-14T18:14:24Z: alacritty::term: scroll_down_relative: origin=0, lines=1
TRACE 2018-03-14T18:14:24Z: alacritty::term: scroll_down_relative: origin=0, lines=1
TRACE 2018-03-14T18:14:24Z: alacritty::term: scroll_down_relative: origin=0, lines=1
TRACE 2018-03-14T18:14:24Z: alacritty::term: scroll_down_relative: origin=0, lines=1
```

**Comment by jwilm** [comment]
My guess is that this is a problem with _winit_. To confirm, would one of you that are able to reproduce this mind running `alacritty --print-events | grep MouseWheel` and uploading the output here?

**Comment by maximbaz** [comment]
When I scroll in Chrome (or anywhere really), alacritty is printing these lines:

```
glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(16))), event: MouseWheel { delta: LineDelta(1.2171053, 13.996711) } }
glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(16))), event: MouseWheel { delta: LineDelta(1.2171053, 18.256578) } }
glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(16))), event: MouseWheel { delta: LineDelta(1.8256578, 18.256578) } }
glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(16))), event: MouseWheel { delta: LineDelta(0.60855263, 15.822369) } }
glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(16))), event: MouseWheel { delta: LineDelta(0.60855263, 20.082237) } }
glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(16))), event: MouseWheel { delta: LineDelta(1.2171053, 17.648026) } }
glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(16))), event: MouseWheel { delta: LineDelta(0.0, 13.996711) } }
glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(16))), event: MouseWheel { delta: LineDelta(0.0, 13.996711) } }
glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(16))), event: MouseWheel { delta: LineDelta(0.60855263, 15.822369) } }
glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(16))), event: MouseWheel { delta: LineDelta(0.60855263, 13.388158) } }
glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(16))), event: MouseWheel { delta: LineDelta(-0.60855263, 11.5625) } }
glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(16))), event: MouseWheel { delta: LineDelta(0.0, 9.128289) } }
glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(16))), event: MouseWheel { delta: LineDelta(0.0, 11.5625) } }
glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(16))), event: MouseWheel { delta: LineDelta(0.0, 10.953947) } }
glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(16))), event: MouseWheel { delta: LineDelta(1.2171053, 11.5625) } }
glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(16))), event: MouseWheel { delta: LineDelta(0.60855263, 4.868421) } }
glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(16))), event: MouseWheel { delta: LineDelta(1.8256578, 9.128289) } }
glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(16))), event: MouseWheel { delta: LineDelta(1.2171053, 4.2598686) } }
glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(16))), event: MouseWheel { delta: LineDelta(0.0, 0.60855263) } }
```

As soon as I come back to alacritty and make a slightest scroll, I get this:

```
glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914563))), event: MouseWheel { device_id: DeviceId(X(DeviceId(2))), delta: LineDelta(0.0, -55.86513), phase: Moved, modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } }
glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(16))), event: MouseWheel { delta: LineDelta(0.0, 0.60855263) } }
glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914563))), event: MouseWheel { device_id: DeviceId(X(DeviceId(2))), delta: LineDelta(0.0, -0.040570177), phase: Moved, modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } }
glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(16))), event: MouseWheel { delta: LineDelta(0.0, 0.60855263) } }
glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914563))), event: MouseWheel { device_id: DeviceId(X(DeviceId(2))), delta: LineDelta(0.0, -0.040570177), phase: Moved, modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } }
```

Notice the large delta in the first event: `delta: LineDelta(0.0, -55.86513)`

**Comment by jwilm** [comment]
Thanks @maximbaz 

It looks like [winit compares the current position with the last recorded position](https://github.com/tomaka/winit/blob/51181b4347aed6f7c1a384c6810635bd6801b369/src/platform/linux/x11/mod.rs#L609) to get the delta. The problem is that the current position is not updated when a DeviceEvent is received -- only WindowEvents.

**Comment by francesca64** [comment]
I'm pretty sure I've found the fix for this: tomaka/winit#514

If anyone wants to test it out, they can: https://github.com/francesca64/alacritty/tree/scroll-debug

(If not, this will just silently appear in winit 0.14.1 soon enough.)

**Comment by chrisduerr** [comment]
That's amazing! Thanks for all your work @francesca64!

**Comment by synaptiko** [comment]
@francesca64 definitely thank you for the work on this.

I installed your branch but unfortunately I still see it happening. Here's what I'm doing (on i3):
1. have Chromium with this issue opened, scrolled at the beginning the page (on workspace 1)
2. have Alacritty opened with very long file and I'm somewhere in the middle of it (workspace 2)
3. both Chromium and Alacritty are focused; I switch to workspace 1 and I scroll to the bottom of the page (on the last comment here)
4. I switch to the workspace 2 and I just touch the middle wheel to generate one event => it will be scrolling much longer then it should

Here's the log:
```
[ResetScroll] SourceID: 2, DeviceID: 2, SourceExists: true, DeviceExists: true
[Unhandled CSI] action='t', args=[22, 0, 0], intermediates=[]
[Unhandled CSI] action='l', args=[1005], intermediates=[63]
[ScrollEvent] Delta: 87.0, Increment: 15.0, NewPosition: 31770.0, OldPosition: 30465.0
[ResetScroll] SourceID: 2, DeviceID: 2, SourceExists: true, DeviceExists: true
```

Btw. I noticed that this won't happen in case I blur and focus the alacritty (ie. if I have two alacritty window opened next to each other). In this case I see following in the log:
```
[ResetScroll] SourceID: 2, DeviceID: 2, SourceExists: true, DeviceExists: true
[ResetScroll] SourceID: 14, DeviceID: 2, SourceExists: true, DeviceExists: true
```

**Comment by francesca64** [comment]
Yeah, the `Cargo.lock` was using an older commit. I changed my Alacritty branch to point to the latest one now, though if you already have a checkout then `cargo update` is going to be easier than pulling that.

**Comment by synaptiko** [comment]
@francesca64 now it works! :-) I'm not familiar with rust so I didn't know about `cargo update` :-)

Once again thanks for tracking it down and fixing it! Really appreciated :-)

**Comment by francesca64** [comment]
@synaptiko no problem!

@chrisduerr I've figured out how to get IME/dead keys/etc. working on macOS, so you can thank me more once that's done.

**Comment by chrisduerr** [comment]
I will! :D

**Comment by benliepert** [comment]
I am having this behavior in alacritty 0.7.2:
(from $ alacritty --print-events | grep MouseWheel)
```
. . .
[2021-03-15 14:18:11.862878511] [INFO ] [alacritty] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(12))), event: MouseWheel { delta: LineDelta(0.0, -15.0) } }
[2021-03-15 14:18:11.873569660] [INFO ] [alacritty] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(12))), event: MouseWheel { delta: LineDelta(0.0, -15.0) } }
[2021-03-15 14:18:11.886251797] [INFO ] [alacritty] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(12))), event: MouseWheel { delta: LineDelta(0.0, -15.0) } }
[2021-03-15 14:18:11.914556909] [INFO ] [alacritty] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(12))), event: MouseWheel { delta: LineDelta(0.0, -15.0) } }
[2021-03-15 14:18:14.390498883] [INFO ] [alacritty] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(12))), event: MouseWheel { delta: LineDelta(0.0, -15.0) } }
[2021-03-15 14:18:14.390534099] [INFO ] [alacritty] glutin event: WindowEvent { window_id: WindowId(X(WindowId(85983239))), event: MouseWheel { device_id: DeviceId(X(DeviceId(2))), delta: LineDelta(0.0, 79.0), phase: Moved, modifiers: (empty) } }
```
In i3-gaps on arch linux (virtualbox): open alacritty and firefox. scrolling in firefox produces the `DeviceEvent`s, a single scroll tic when returning to alacritty produces the `WindowEvent`, which appears to instantaneously execute 79 scroll events (tics of the mouse wheel). 
I've confirmed this behavior does not happen in gnome-terminal.

**Comment by vinnydiehl** [comment]
I am having this behavior in Alacritty 0.12.1 on Arch + awesome. If I for example go to my web browser and scroll up a ton, I can then go around to every single Alacritty window and scroll up one click to send each window waaaaaaay up the scrollback.

When watching `alacritty --print-events | grep MouseWheel`, only one winit DeviceEvent is sent when the bug occurs.

@benliepert is this still an issue for you? @jwilm any thoughts?

**Comment by kchibisov** [comment]
>I am having this behavior in Alacritty 0.12.1 on Arch + awesome. If I for example go to my web browser and scroll up a ton, I can then go around to every single Alacritty window and scroll up one click to send each window waaaaaaay up the scrollback.

If you have an issue like that open a new issue instead of bumping old ones. I'd suspect that the delta values in `MouseWheel` events are too high.

> you? @jwilm any thoughts?

Would suggest to not do arbitrary pings on 5 years old issues.
